
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Reproducible bioinformatics toolkit with schema-verified visualization">
      
      
      
        <link rel="canonical" href="https://omniscoder.github.io/Helix/reference/api/">
      
      
        <link rel="prev" href="../../cli/demo/">
      
      
        <link rel="next" href="../../whats-new/">
      
      
        
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>helix.api - Veri-Helix</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#helixapi" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Veri-Helix" class="md-header__button md-logo" aria-label="Veri-Helix" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Veri-Helix
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              helix.api
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="deep-orange"  aria-hidden="true"  type="radio" name="__palette" id="__palette_0">
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/omniscoder/Helix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    omniscoder/Helix
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Veri-Helix" class="md-nav__button md-logo" aria-label="Veri-Helix" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Veri-Helix
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/omniscoder/Helix" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    omniscoder/Helix
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../getting-started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Getting Started
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Concepts
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Concepts
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/artifacts/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Artifacts & Provenance
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/schemas/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schemas
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Workflows
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    CLI
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    CLI
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/viz/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Visualization
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/schema/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema Tools
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Workflows
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demos
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    API Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    helix.api
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    helix.api
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#function-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Function reference
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Function reference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna_summarysequencenone-input_pathnone-window200-step50-k5-max_diff1" class="md-nav__link">
    <span class="md-ellipsis">
      
        dna_summary(sequence=None, *, input_path=None, window=200, step=50, k=5, max_diff=1)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#triage_reportsequencenone-input_pathnone-k5-max_diff1-min_orf_length90" class="md-nav__link">
    <span class="md-ellipsis">
      
        triage_report(sequence=None, *, input_path=None, k=5, max_diff=1, min_orf_length=90)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fold_rnasequence-min_loop_length3-allow_wobble_pairstrue" class="md-nav__link">
    <span class="md-ellipsis">
      
        fold_rna(sequence, *, min_loop_length=3, allow_wobble_pairs=True)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#spectrum_leaderboardpeptidenone-experimental_spectrumnone-cyclictrue-leaderboard_size5" class="md-nav__link">
    <span class="md-ellipsis">
      
        spectrum_leaderboard(peptide=None, *, experimental_spectrum=None, cyclic=True, leaderboard_size=5)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#protein_summarysequencenone-input_pathnone-window9-step1-scalekd" class="md-nav__link">
    <span class="md-ellipsis">
      
        protein_summary(sequence=None, *, input_path=None, window=9, step=1, scale="kd")
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run_workflowconfig_path-output_dir-namenone" class="md-nav__link">
    <span class="md-ellipsis">
      
        run_workflow(config_path, *, output_dir, name=None)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-patterns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Usage patterns
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../whats-new/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    What's New
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Visualization Gallery
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    Visualization Gallery
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/alignment-ribbon/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Alignment Ribbon
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/rna-dotplot/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA Dot-Plot
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/motif-logo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Motif Logo
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/distance-heatmap/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Distance Heatmap
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/minimizer-density/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Minimizer Density
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../schema-reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema Reference
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../benchmarks/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Benchmarks
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Contributing
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#function-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Function reference
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Function reference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna_summarysequencenone-input_pathnone-window200-step50-k5-max_diff1" class="md-nav__link">
    <span class="md-ellipsis">
      
        dna_summary(sequence=None, *, input_path=None, window=200, step=50, k=5, max_diff=1)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#triage_reportsequencenone-input_pathnone-k5-max_diff1-min_orf_length90" class="md-nav__link">
    <span class="md-ellipsis">
      
        triage_report(sequence=None, *, input_path=None, k=5, max_diff=1, min_orf_length=90)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fold_rnasequence-min_loop_length3-allow_wobble_pairstrue" class="md-nav__link">
    <span class="md-ellipsis">
      
        fold_rna(sequence, *, min_loop_length=3, allow_wobble_pairs=True)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#spectrum_leaderboardpeptidenone-experimental_spectrumnone-cyclictrue-leaderboard_size5" class="md-nav__link">
    <span class="md-ellipsis">
      
        spectrum_leaderboard(peptide=None, *, experimental_spectrum=None, cyclic=True, leaderboard_size=5)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#protein_summarysequencenone-input_pathnone-window9-step1-scalekd" class="md-nav__link">
    <span class="md-ellipsis">
      
        protein_summary(sequence=None, *, input_path=None, window=9, step=1, scale="kd")
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run_workflowconfig_path-output_dir-namenone" class="md-nav__link">
    <span class="md-ellipsis">
      
        run_workflow(config_path, *, output_dir, name=None)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-patterns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Usage patterns
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/omniscoder/Helix/edit/master/docs/reference/api.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="helixapi"><code>helix.api</code><a class="headerlink" href="#helixapi" title="Permanent link">&para;</a></h1>
<p>Research workflows often start in notebooks or lightweight scripts where shelling out to the CLI is inconvenient. The <code>helix.api</code> module mirrors the CLI surface area but returns plain Python dictionaries and lists, making it straightforward to serialize to JSON, pass through pandas, or feed into downstream visualization components.</p>
<blockquote>
<p><strong>Simulation only:</strong> <code>helix.api</code> helpers work exclusively on digital sequences, spectra, or graphs. They never prescribe wet-lab procedures or interface with instruments—use them strictly for in-silico analysis and visualization.</p>
</blockquote>
<p>The helpers prefer explicit inputs: every function accepts either an inline <code>sequence="ACGU..."</code> <strong>or</strong> an <code>input_path=Path("sample.fasta")</code> when DNA/protein IO is relevant. Validation happens up front—invalid bases, overlapping arguments, and missing files raise informative <code>ValueError</code> or <code>ImportError</code> exceptions so that provenance remains audit-friendly.</p>
<h2 id="function-reference">Function reference<a class="headerlink" href="#function-reference" title="Permanent link">&para;</a></h2>
<h3 id="dna_summarysequencenone-input_pathnone-window200-step50-k5-max_diff1"><code>dna_summary(sequence=None, *, input_path=None, window=200, step=50, k=5, max_diff=1)</code><a class="headerlink" href="#dna_summarysequencenone-input_pathnone-window200-step50-k5-max_diff1" title="Permanent link">&para;</a></h3>
<p>Normalize a DNA string, compute GC statistics, and discover k-mer hotspots tolerant of SNP-scale variation.</p>
<p><strong>Parameters</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sequence</code></td>
<td><code>str \| None</code></td>
<td>Inline DNA input. Cannot be combined with <code>input_path</code>.</td>
</tr>
<tr>
<td><code>input_path</code></td>
<td><code>str \| Path \| None</code></td>
<td>File to read (FASTA/plain). Takes precedence over <code>sequence</code>.</td>
</tr>
<tr>
<td><code>window</code></td>
<td><code>int</code></td>
<td>Sliding-window size for GC summaries. Set to <code>0</code> to skip.</td>
</tr>
<tr>
<td><code>step</code></td>
<td><code>int</code></td>
<td>Advance between windows; smaller steps increase resolution.</td>
</tr>
<tr>
<td><code>k</code></td>
<td><code>int</code></td>
<td>k-mer size used for clustering recurrent motifs.</td>
</tr>
<tr>
<td><code>max_diff</code></td>
<td><code>int</code></td>
<td>Maximum Hamming distance when grouping similar k-mers.</td>
</tr>
</tbody>
</table>
<p><strong>Returns</strong></p>
<p><code>dict</code> with:</p>
<ul>
<li><code>sequence</code>: normalized uppercase DNA (U is converted to T, FASTA headers stripped).  </li>
<li><code>length</code>: integer length of <code>sequence</code>.  </li>
<li><code>gc_content</code>: float fraction (0–1).  </li>
<li><code>gc_windows</code>: list of <code>{start, end, gc_fraction}</code> windows for downstream plotting.  </li>
<li><code>kmer_clusters</code>: map keyed by canonical motif with <code>count</code>, <code>positions</code>, <code>patterns</code>.  </li>
</ul>
<p>The GC window calculation and k-mer clustering match the logic behind <code>helix dna summarize</code>, ensuring CLI and notebook reports stay interchangeable.</p>
<hr />
<h3 id="triage_reportsequencenone-input_pathnone-k5-max_diff1-min_orf_length90"><code>triage_report(sequence=None, *, input_path=None, k=5, max_diff=1, min_orf_length=90)</code><a class="headerlink" href="#triage_reportsequencenone-input_pathnone-k5-max_diff1-min_orf_length90" title="Permanent link">&para;</a></h3>
<p>Produce the complete “triage” bundle (GC skew, motif clusters, ORF calls) that backs <code>helix viz triage</code>.</p>
<p><strong>Parameters</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sequence</code>, <code>input_path</code></td>
<td>see above</td>
<td>Mutually exclusive DNA inputs.</td>
</tr>
<tr>
<td><code>k</code>, <code>max_diff</code></td>
<td><code>int</code></td>
<td>Passed through to the underlying k-mer clustering.</td>
</tr>
<tr>
<td><code>min_orf_length</code></td>
<td><code>int</code></td>
<td>Filter ORFs shorter than the threshold (nt).</td>
</tr>
</tbody>
</table>
<p><strong>Returns</strong></p>
<p><code>dict</code> with:</p>
<ul>
<li><code>sequence</code>: RNA display sequence (T→U) for visualization.  </li>
<li><code>skew</code>: list of cumulative skew values (len = sequence+1).  </li>
<li><code>clusters</code>: each cluster has <code>canonical</code>, <code>count</code>, <code>patterns</code>, <code>positions</code>.  </li>
<li><code>orfs</code>: list of ORF dicts reporting <code>start</code>, <code>end</code>, <code>strand</code>, <code>frame</code>, <code>length_nt</code>, <code>length_aa</code>, and translated <code>peptide</code>.  </li>
</ul>
<p>All payloads are schema-compatible with <code>helix schema show triage-report</code>.</p>
<hr />
<h3 id="fold_rnasequence-min_loop_length3-allow_wobble_pairstrue"><code>fold_rna(sequence, *, min_loop_length=3, allow_wobble_pairs=True)</code><a class="headerlink" href="#fold_rnasequence-min_loop_length3-allow_wobble_pairstrue" title="Permanent link">&para;</a></h3>
<p>Convenience wrapper around the annotated Nussinov dynamic program for RNA folding.</p>
<p><strong>Parameters</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sequence</code></td>
<td><code>str</code></td>
<td>RNA/DNA string; U/T are normalized to U internally.</td>
</tr>
<tr>
<td><code>min_loop_length</code></td>
<td><code>int</code></td>
<td>Nussinov “hairpin” constraint (nt separating paired bases).</td>
</tr>
<tr>
<td><code>allow_wobble_pairs</code></td>
<td><code>bool</code></td>
<td>When <code>True</code>, GU wobble pairs are permitted in addition to AU/GC.</td>
</tr>
</tbody>
</table>
<p><strong>Returns</strong></p>
<p><code>dict</code> with:</p>
<ul>
<li><code>sequence</code>: normalized RNA string used for folding.  </li>
<li><code>score</code>: optimal base-pair count reported by the DP table.  </li>
<li><code>pairs</code>: list of <code>(i, j)</code> tuples describing paired indices.  </li>
<li><code>dot_bracket</code>: canonical dot-bracket representation for plotting or downstream structure comparison.  </li>
</ul>
<p>Because the helper exposes the same knobs as <code>helix rna fold</code>, results are reproducible across interfaces.</p>
<hr />
<h3 id="spectrum_leaderboardpeptidenone-experimental_spectrumnone-cyclictrue-leaderboard_size5"><code>spectrum_leaderboard(peptide=None, *, experimental_spectrum=None, cyclic=True, leaderboard_size=5)</code><a class="headerlink" href="#spectrum_leaderboardpeptidenone-experimental_spectrumnone-cyclictrue-leaderboard_size5" title="Permanent link">&para;</a></h3>
<p>Run the leaderboard cyclopeptide sequencing algorithm and return notebook-friendly results.</p>
<p><strong>Parameters</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>peptide</code></td>
<td><code>str \| None</code></td>
<td>Optional candidate peptide for generating a theoretical spectrum.</td>
</tr>
<tr>
<td><code>experimental_spectrum</code></td>
<td><code>Sequence[int] \| None</code></td>
<td>Observed masses used for leaderboard selection.</td>
</tr>
<tr>
<td><code>cyclic</code></td>
<td><code>bool</code></td>
<td>Whether the theoretical spectrum should be cyclic (<code>True</code>) or linear (<code>False</code>).</td>
</tr>
<tr>
<td><code>leaderboard_size</code></td>
<td><code>int</code></td>
<td>Maximum number of peptides kept per iteration (ties are preserved).</td>
</tr>
</tbody>
</table>
<p><strong>Returns</strong></p>
<p><code>dict</code> with:</p>
<ul>
<li><code>theoretical_spectrum</code>: the computed linear/cyclic spectrum for <code>peptide</code> (empty if no peptide supplied).  </li>
<li><code>leaderboard_hits</code>: list of <code>{peptide, score}</code> pairs sorted by score.  </li>
</ul>
<p>Pass only <code>peptide</code> to preview its spectrum, only <code>experimental_spectrum</code> to search for best-scoring sequences, or both to compare expectations vs. observed data.</p>
<hr />
<h3 id="protein_summarysequencenone-input_pathnone-window9-step1-scalekd"><code>protein_summary(sequence=None, *, input_path=None, window=9, step=1, scale="kd")</code><a class="headerlink" href="#protein_summarysequencenone-input_pathnone-window9-step1-scalekd" title="Permanent link">&para;</a></h3>
<p>Summarize amino-acid sequences using Biopython’s <code>ProtParam</code> utilities along with Helix hydropathy profiles.</p>
<p><strong>Parameters</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sequence</code>, <code>input_path</code></td>
<td>see above</td>
<td>FASTA headers are handled transparently.</td>
</tr>
<tr>
<td><code>window</code></td>
<td><code>int</code></td>
<td>Sliding window for hydropathy averaging.</td>
</tr>
<tr>
<td><code>step</code></td>
<td><code>int</code></td>
<td>Offset between successive hydropathy windows.</td>
</tr>
<tr>
<td><code>scale</code></td>
<td><code>str</code></td>
<td>Hydropathy scale identifier (e.g., <code>"kd"</code> for Kyte-Doolittle).</td>
</tr>
</tbody>
</table>
<p><strong>Returns</strong></p>
<p><code>dict</code> with canonical protein metrics: <code>sequence</code>, <code>length</code>, <code>molecular_weight</code>, <code>aromaticity</code>, <code>instability_index</code>, <code>gravy</code>, <code>charge_at_pH7</code>, plus <code>hydropathy_profile</code> entries (<code>start</code>, <code>end</code>, <code>score</code>). Raises <code>ImportError</code> if Biopython is unavailable so that workflows can fail fast with a clear dependency message.</p>
<hr />
<h3 id="run_workflowconfig_path-output_dir-namenone"><code>run_workflow(config_path, *, output_dir, name=None)</code><a class="headerlink" href="#run_workflowconfig_path-output_dir-namenone" title="Permanent link">&para;</a></h3>
<p>Execute a YAML workflow definition (the same format consumed by <code>helix workflows run</code>) from Python. The helper wraps <code>helix_workflows.run_workflow_config</code>, returning the list of materialized artifacts. <code>name</code> can restrict execution to a single workflow from a manifest, mirroring the CLI <code>--name</code> flag.</p>
<h2 id="usage-patterns">Usage patterns<a class="headerlink" href="#usage-patterns" title="Permanent link">&para;</a></h2>
<div class="codehilite"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">helix</span><span class="w"> </span><span class="kn">import</span> <span class="n">api</span> <span class="k">as</span> <span class="n">hx</span>

<span class="n">report</span> <span class="o">=</span> <span class="n">hx</span><span class="o">.</span><span class="n">triage_report</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;AUGGCCUUUUAA&quot;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">gc_bins</span> <span class="o">=</span> <span class="n">hx</span><span class="o">.</span><span class="n">dna_summary</span><span class="p">(</span><span class="n">input_path</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;samples/ecoli.fna&quot;</span><span class="p">),</span> <span class="n">window</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">rna</span> <span class="o">=</span> <span class="n">hx</span><span class="o">.</span><span class="n">fold_rna</span><span class="p">(</span><span class="s2">&quot;GGGAAACCC&quot;</span><span class="p">,</span> <span class="n">min_loop_length</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">peptides</span> <span class="o">=</span> <span class="n">hx</span><span class="o">.</span><span class="n">spectrum_leaderboard</span><span class="p">(</span>
    <span class="n">experimental_spectrum</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">113</span><span class="p">,</span> <span class="mi">128</span><span class="p">,</span> <span class="mi">227</span><span class="p">,</span> <span class="mi">242</span><span class="p">,</span> <span class="mi">355</span><span class="p">,</span> <span class="mi">370</span><span class="p">,</span> <span class="mi">484</span><span class="p">],</span>
    <span class="n">leaderboard_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>

<p>Each function returns plain JSON-serializable structures, making it trivial to call <code>json.dumps(...)</code>, ship results to <code>helix viz ...</code>, or interoperate with scientific Python stacks. Exceptions surface early and with actionable messages—ideal for research-grade reproducibility and provenance.</p>
<p>Need performance baselines? Run <code>python -m benchmarks.api_benchmarks --repeat 5 --limit 0 --sort mean --out bench/api.json</code> from the repo root to capture timing data for every helper (or focus on specific functions via <code>--scenario</code>). Set <code>HELIX_BENCH_DNA_FASTA</code> / <code>HELIX_BENCH_PROTEIN_FASTA</code> to swap in larger genomes or proteomes once available, and use <code>--limit 10000</code> to mimic CI’s quicker sampling. Each run emits a schema-tagged payload (<code>bench_result</code> v1.0) that logs git SHA, BLAS vendor, CPU/threads, RNG seed, and per-case RSS stats so notebooks + CI can do apples-to-apples comparisons. Compare two runs via <code>scripts/bench_check.py baseline.json current.json --threshold 5</code> to flag &gt;5% slowdowns automatically, and browse the rolling history at <code>docs/benchmarks.md</code>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["navigation.instant", "navigation.sections", "content.code.copy", "content.action.edit"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>